src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/String/KeyAnalyzer.cs

```diff
@@ -263,25 +245,34 @@ public AnalysisResults(bool ignoreCase, bool allAsciiIfIgnoreCase, int hashIndex
             public bool RightJustifiedSubstring => HashIndex < 0;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ReadOnlySpan<char> Slicer(this string s, int index, int count) => s.AsSpan((index >= 0 ? index : s.Length + index), count);
+
         private abstract class SubstringComparer : IEqualityComparer<string>
         {
-            public int Index;
-            public int Count;
-            public bool IsLeft;
+            public int Index;   // offset from left side (if positive) or right side (if negative) of the string
+            public int Count;   // number of characters in the span
+
             public abstract bool Equals(string? x, string? y);
             public abstract int GetHashCode(string s);
         }
 
         private sealed class JustifiedSubstringComparer : SubstringComparer
         {
-            public override bool Equals(string? x, string? y) => x.AsSpan(IsLeft ? Index : (x!.Length + Index), Count).SequenceEqual(y.AsSpan(IsLeft ? Index : (y!.Length + Index), Count));
-            public override int GetHashCode(string s) => Hashing.GetHashCodeOrdinal(s.AsSpan(IsLeft ? Index : (s.Length + Index), Count));
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            public override bool Equals(string? x, string? y) => x!.Slicer(Index, Count).SequenceEqual(y!.Slicer(Index, Count));
```

Comment:
The Slicer does the left/right based on the sign of the `Index` value now, which should JIT down better. It might be fun to figure out if there's a better way to do this as inline math instead of a ternary, but `String.AsSpan` doesn't work in a way that would make this easy.